verification/cbmc/include/core_json_annex.h

```diff
@@ -67,45 +67,71 @@
  * Please see core_json.c for documentation.
  */
 
-void skipSpace( const char * buf,
-                size_t * start,
-                size_t max );
-
-bool skipUTF8( const char * buf,
-               size_t * start,
-               size_t max );
-
-bool skipEscape( const char * buf,
-                 size_t * start,
-                 size_t max );
-
-bool skipString( const char * buf,
-                 size_t * start,
-                 size_t max );
-
-bool skipAnyLiteral( const char * buf,
-                     size_t * start,
-                     size_t max );
-
-bool skipDigits( const char * buf,
-                 size_t * start,
-                 size_t max,
-                 int32_t * outValue );
+/* *INDENT-OFF* */
+JSONStatus_t JSON_Validate( const char * buf,
+                            size_t max )
+__CPROVER_requires( max < CBMC_MAX_BUFSIZE )
+__CPROVER_requires( buf == NULL || __CPROVER_is_fresh( buf, max ) )
+__CPROVER_ensures( jsonValidateEnum( __CPROVER_return_value ) );
 
-bool skipNumber( const char * buf,
-                 size_t * start,
-                 size_t max );
-
-bool skipSpaceAndComma( const char * buf,
-                        size_t * start,
-                        size_t max );
+JSONStatus_t skipCollection( const char * buf,
+                             size_t * start,
+                             size_t max )
+__CPROVER_requires( max > 0 && max < CBMC_MAX_BUFSIZE )
+__CPROVER_requires( __CPROVER_is_fresh( buf, max ) )
+__CPROVER_requires( __CPROVER_is_fresh( start, sizeof( *start ) ) )
+__CPROVER_assigns( *start )
+__CPROVER_ensures( skipCollectionEnum( __CPROVER_return_value ) )
+__CPROVER_ensures( __CPROVER_return_value == JSONSuccess ==> *start <= max );
+
+void skipScalars( const char * buf,
+                  size_t * start,
+                  size_t max,
+                  char mode )
+__CPROVER_requires( max > 0 && max < CBMC_MAX_BUFSIZE )
+__CPROVER_requires( __CPROVER_is_fresh( buf, max ) )
+__CPROVER_requires( __CPROVER_is_fresh( start, sizeof( *start ) ) )
+__CPROVER_requires( ( mode == '{' ) || ( mode == '[' ) )
+/* __CPROVER_requires(isOpenBracket_( mode )) */
+__CPROVER_assigns( *start )
+__CPROVER_ensures( *start >= __CPROVER_old( *start ) )
+/* __CPROVER_ensures(__CPROVER_old(*start) <= max ==> *start <= max) */
+__CPROVER_ensures( __CPROVER_old( *start ) > max ==> *start == __CPROVER_old( *start ) );
 
 bool skipAnyScalar( const char * buf,
                     size_t * start,
-                    size_t max );
+                    size_t max )
+__CPROVER_requires( max > 0 && max < CBMC_MAX_BUFSIZE )
+__CPROVER_requires( __CPROVER_is_fresh( buf, max ) )
+__CPROVER_requires( __CPROVER_is_fresh( start, sizeof( *start ) ) )
+__CPROVER_assigns( *start )
```

Comment:
Check with Remi, but I think we have conditional assigns clauses now, which would let you say that *start is only assigned if *start < max, which would be more compact, and perhaps also improve proof performance.

Reply:
Unless the gain is significant, I'd prefer to keep the assigns clauses as simple as they are now. @remi-delmas-3000, any recommendations?

Reply:
with a conditional target you'd write this and just get rid of the last ensures clause:

```c
__CPROVER_requires( max > 0 && max < CBMC_MAX_BUFSIZE )
__CPROVER_requires( __CPROVER_is_fresh( buf, max ) )
__CPROVER_requires( __CPROVER_is_fresh( start, sizeof( *start ) ) )
__CPROVER_assigns( *start <= max : *start )
__CPROVER_ensures( isBool( __CPROVER_return_value ) )
__CPROVER_ensures( *start >= __CPROVER_old( *start ) )
__CPROVER_ensures( __CPROVER_old( *start ) <= max ==> *start <= max )
```